File: src/main/core/io/GlobalPath.ts
import * as path from "path";
import * as os from "os";
import Utils from "../utils/Utils";

export default class GlobalPath {

  public static getGameDataDirPath(): string {

    if(Utils.getOSType() === "windows") {

      return path.join(os.homedir(), "AppData", "Roaming", ".mckismetlab");

    } else if(Utils.getOSType() === "osx") {

      return path.join(os.homedir() + "/Library/Application Support/mckismetlab");

    } else {

      throw new Error("unknown os type.");

    }

  }

  public static getInstancesDirPath(): string {
    return path.join(this.getGameDataDirPath(), "instances");
  }

  public static getCommonDirPath(): string {
    return path.join(this.getGameDataDirPath(), "common");
  }
}

File: namelessrealms_core/src/util/global_path.rs
use std::path::{PathBuf, Path};

use super::{config, utils::{OSType, self}};

pub fn get_app_dir_path() -> PathBuf {

    let app_dir_name = config::APP_DIR_NAME;
    let home_dir = home::home_dir().unwrap();

    match utils::get_os_type() {
        OSType::Windows =>  home_dir.join("AppData").join("Roaming").join(app_dir_name),
        OSType::MacOS => home_dir.join("Library").join("Application Support").join(app_dir_name),
        OSType::Linux => home_dir.join(app_dir_name)
    }
}

pub fn get_instances_dir_path() -> PathBuf {
    self::get_app_dir_path().join("instances")
}

pub fn get_common_dir_path() -> PathBuf {
    self::get_app_dir_path().join("common")
}

// pub fn get_logs_dir_path() -> PathBuf {
//     get_app_dir_path().join("logs")
// }

pub fn combine_common_paths_absolute(absolute_dir: &Path, path: &Path) -> PathBuf {
    self::get_common_dir_path().join(absolute_dir).join(path)
}

pub fn get_caches_dir_path() -> PathBuf {
    self::get_app_dir_path().join("caches")
}

pub fn get_processes_json_file_path() -> PathBuf {
    self::get_caches_dir_path().join("processes.json")
}